app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaAppState.kt

```diff
@@ -0,0 +1,131 @@
+import androidx.navigation.NavController
+import androidx.navigation.NavDestination
+import androidx.navigation.NavGraph.Companion.findStartDestination
+import androidx.navigation.NavHostController
+import androidx.navigation.compose.currentBackStackEntryAsState
+import androidx.navigation.compose.rememberNavController
+import androidx.tracing.trace
+import com.google.samples.apps.nowinandroid.core.designsystem.icon.Icon.DrawableResourceIcon
+import com.google.samples.apps.nowinandroid.core.designsystem.icon.Icon.ImageVectorIcon
+import com.google.samples.apps.nowinandroid.core.designsystem.icon.NiaIcons
+import com.google.samples.apps.nowinandroid.core.navigation.NiaNavigationDestination
+import com.google.samples.apps.nowinandroid.core.ui.JankMetricDisposableEffect
+import com.google.samples.apps.nowinandroid.feature.foryou.R as forYouR
+import com.google.samples.apps.nowinandroid.feature.foryou.navigation.ForYouDestination
+import com.google.samples.apps.nowinandroid.feature.interests.R as interestsR
+import com.google.samples.apps.nowinandroid.feature.interests.navigation.InterestsDestination
+import com.google.samples.apps.nowinandroid.navigation.TopLevelDestination
+
+@Composable
+fun rememberNiaAppState(
+    windowSizeClass: WindowSizeClass,
+    navController: NavHostController = rememberNavController()
+): NiaAppState {
+    NavigationTrackingSideEffect(navController)
+    return remember(navController, windowSizeClass) {
+        NiaAppState(navController, windowSizeClass)
+    }
+}
+
+@Stable
+class NiaAppState(
+    val navController: NavHostController,
+    val windowSizeClass: WindowSizeClass
+) {
+    val currentDestination: NavDestination?
+        @Composable get() = navController
+            .currentBackStackEntryAsState().value?.destination
+
+    val shouldShowBottomBar: Boolean
+        get() = windowSizeClass.widthSizeClass == WindowWidthSizeClass.Compact ||
+            windowSizeClass.heightSizeClass == WindowHeightSizeClass.Compact
+
+    val shouldShowNavRail: Boolean
+        get() = !shouldShowBottomBar
+
+    val topLevelDestinations: List<TopLevelDestination> = listOf(
+        TopLevelDestination(
+            route = ForYouDestination.route,
+            destination = ForYouDestination.destination,
+            selectedIcon = DrawableResourceIcon(NiaIcons.Upcoming),
+            unselectedIcon = DrawableResourceIcon(NiaIcons.UpcomingBorder),
+            iconTextId = forYouR.string.for_you
+        ),
+        TopLevelDestination(
+            route = InterestsDestination.route,
+            destination = InterestsDestination.destination,
+            selectedIcon = ImageVectorIcon(NiaIcons.Grid3x3),
+            unselectedIcon = ImageVectorIcon(NiaIcons.Grid3x3),
+            iconTextId = interestsR.string.interests
+        )
+    )
+
+    // --------------------
+    // NAVIGATION LOGIC
+    // --------------------
+
+    fun navigate(destination: NiaNavigationDestination, route: String? = null) {
```

Comment:
Can I request a KDoc here? Just looking at the method being called, it's not super clear to me how the `route` and `destination` relate to each other, and the implications of a `route` not being supplied.

Reply:
Sure! Added, thanks